File: Assets/InstallerSource/VrcGetCs/RepoHolder.cs
// ReSharper disable InconsistentNaming

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;
using Anatawa12.SimpleJson;
using JetBrains.Annotations;
using UnityEngine;
using static Anatawa12.VrcGet.ModStatics;

namespace Anatawa12.VrcGet
{
    class RepoHolder
    {
        // the pointer of LocalCachedRepository will never be changed
        [NotNull] readonly Dictionary<Path, LocalCachedRepository> cached_repos_new;

        public RepoHolder()
        {
            cached_repos_new = new Dictionary<Path, LocalCachedRepository>();
        }

        internal async Task load_repos([CanBeNull] HttpClient http, [NotNull] [ItemNotNull] IEnumerable<RepoSource> sources)
        {
            var repos = await Task.WhenAll(sources.Select(async src =>
                (await load_repo_from_source(http, src), src.file_path())));

            foreach (var (repo, path) in repos)
            {
                cached_repos_new[path] = repo;
            }
        }

        static async Task<LocalCachedRepository> load_repo_from_source([CanBeNull] HttpClient client, [NotNull] RepoSource source)
        {
            return await source.VisitLoadRepo(client);
        }

        public static async Task<LocalCachedRepository> LoadPreDefinedRepo([CanBeNull] HttpClient client,
            [NotNull] PreDefinedRepoSource source)
        {
            return await load_remote_repo(
                client,
                null,
                source.path,
                source.Info.url
            );
        }

        public static async Task<LocalCachedRepository> LoadUserRepo([CanBeNull] HttpClient client,
            [NotNull] UserRepoSource source)
        {
            var user_repo = source.Setting;
            if (user_repo.url is string url)
            {
                return await load_remote_repo(
                    client,
                    user_repo.headers,
                    user_repo.local_path,
                    url
                );
            }
            else
            {
                return await load_local_repo(client, user_repo.local_path);
            }
        }

        static async Task<LocalCachedRepository> load_remote_repo(
            [CanBeNull] HttpClient client,
            [CanBeNull] Dictionary<String, String> headers,
            [NotNull] Path path,
            [NotNull] string remote_url
        )
        {
            return await load_repo(path, client, async () =>
            {
                // if local repository not found: try downloading remote one
                if (client == null) throw new OfflineModeException();

                var may_null = await download_remote_repository(client, remote_url, headers, null);
                System.Diagnostics.Debug.Assert(may_null != null, nameof(may_null) + " != null");
                var (remote_repo, etag) = may_null.Value;

                var local_cache = new LocalCachedRepository(remote_repo, headers ?? new Dictionary<string, string>());

                if (etag != null)
                {
                    local_cache.vrc_get = local_cache.vrc_get ?? new VrcGetMeta();
                    local_cache.vrc_get.etag = etag;
                }

                try
                {
                    await write_repo(path, local_cache);
                }
                catch (Exception e)
                {
                    Debug.LogError("Error writing local repository");
                    Debug.LogException(e);
                }

                return local_cache;
            });
        }

        static async Task<LocalCachedRepository> load_local_repo(
            [CanBeNull] HttpClient client,
            [NotNull] Path path
        )
        {
            return await load_repo(path, client, () => throw new IOException("repository not found"));
        }

        static async Task<LocalCachedRepository> load_repo(
            [NotNull] Path path,
            [CanBeNull] HttpClient http,
            Func<Task<LocalCachedRepository>> if_not_found
        )
        {
            string text;
            try
            {
                text = await TryReadFile(path);
            }
            catch
            {
                return await if_not_found();
            }

            if (text == null)
                return await if_not_found();

            var loaded = new LocalCachedRepository(new JsonParser(text).Parse(JsonType.Obj));
            if (http != null)
                await update_from_remote(http, path, loaded);
            return loaded;
        }

        public LocalCachedRepository[] get_repos() => cached_repos_new.Values.ToArray();

        public IEnumerable<(Path, LocalCachedRepository)> get_repo_with_path() =>
            cached_repos_new.Select(x => (x.Key, x.Value));

        [CanBeNull]
        internal LocalCachedRepository get_repo(Path path) => cached_repos_new.get(path);

        // VPAI: to add pending repo
        public void AddRepository(Path path, LocalCachedRepository cache)
        {
            cached_repos_new.Add(path, cache);
        }

        public void remove_repo(Path path)
        {
            cached_repos_new.Remove(path);
        }
    }
}

File: vrc-get/src/vpm/repo_holder.rs
use super::*;
use std::collections::HashMap;
use std::future::Future;

#[derive(Debug)]
pub(super) struct RepoHolder {
    http: Option<Client>,
    cached_repos_new: HashMap<PathBuf, LocalCachedRepository>,
}

impl RepoHolder {
    pub(crate) fn new(http: Option<Client>) -> Self {
        RepoHolder {
            http,
            cached_repos_new: HashMap::new(),
        }
    }
}

// new system
impl RepoHolder {
    pub(crate) async fn load_repos(&mut self, sources: Vec<RepoSource>) -> io::Result<()> {
        fn file_path(source: &RepoSource) -> &Path {
            match source {
                RepoSource::PreDefined(_, path) => path,
                RepoSource::UserRepo(user) => &user.local_path,
                RepoSource::Undefined(undefined) => undefined,
            }
        }

        let repos = try_join_all(sources.iter().map(|src| async {
            Self::load_repo_from_source(self.http.as_ref(), src).await
                .map(|v| (v, file_path(src)))
        }))
        .await?;

        for (repo, path) in repos {
            self.cached_repos_new.insert(path.to_owned(), repo);
        }

        Ok(())
    }

    async fn load_repo_from_source(
        client: Option<&Client>,
        source: &RepoSource,
    ) -> io::Result<LocalCachedRepository> {
        match source {
            RepoSource::PreDefined(source, path) => {
                RepoHolder::load_remote_repo(
                    client,
                    None,
                    &path, 
                    source.url,
                ).await
            }
            RepoSource::UserRepo(user_repo) => {
                if let Some(url) = &user_repo.url {
                    RepoHolder::load_remote_repo(
                        client,
                        Some(&user_repo.headers),
                        &user_repo.local_path,
                        &url,
                    )
                    .await
                } else {
                    RepoHolder::load_local_repo(client, &user_repo.local_path).await
                }
            }
            RepoSource::Undefined(repo_json) => {
                RepoHolder::load_local_repo(client, &repo_json).await
            }
        }
    }

    async fn load_remote_repo(
        client: Option<&Client>,
        headers: Option<&IndexMap<String, String>>,
        path: &Path,
        remote_url: &str,
    ) -> io::Result<LocalCachedRepository> {
        Self::load_repo(path, client, || async {
            // if local repository not found: try downloading remote one
            let Some(client) = client else {
                return Err(io::Error::new(io::ErrorKind::ConnectionAborted, "offline mode"))
            };
            let (remote_repo, etag) = download_remote_repository(&client, remote_url, headers, None)
                .await?
                .expect("logic failure: no etag");

            let mut local_cache = LocalCachedRepository::new(remote_repo, headers.map(Clone::clone).unwrap_or_default());

            if let Some(etag) = etag {
                local_cache
                    .vrc_get
                    .get_or_insert_with(Default::default)
                    .etag = etag;
            }

            match write_repo(path, &local_cache).await {
                Ok(_) => {}
                Err(e) => {
                    log::error!("writing local repo '{}': {}", path.display(), e);
                }
            }

            Ok(local_cache)
        })
        .await
    }

    async fn load_local_repo(
        client: Option<&Client>,
        path: &Path,
    ) -> io::Result<LocalCachedRepository> {
        Self::load_repo(path, client, || async { unreachable!() }).await
    }

    async fn load_repo<F, T>(
        path: &Path,
        http: Option<&Client>,
        if_not_found: F,
    ) -> io::Result<LocalCachedRepository>
    where
        F: FnOnce() -> T,
        T: Future<Output = io::Result<LocalCachedRepository>>,
    {
        let Some(json_file) = try_open_file(path).await? else {
            return Ok(if_not_found().await?);
        };

        let mut loaded = match serde_json::from_slice(&read_to_vec(json_file).await?) {
            Ok(loaded) => loaded,
            Err(e) => {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    format!("loading {}: {}", path.display(), e),
                ))
            }
        };
        if let Some(http) = http {
            update_from_remote(http, path.into(), &mut loaded).await;
        }
        return Ok(loaded);
    }

    pub(crate) fn get_repos(&self) -> Vec<&LocalCachedRepository> {
        self.cached_repos_new.values().collect()
    }

    pub(crate) fn get_repo_with_path(&self) -> impl Iterator<Item = (&'_ PathBuf, &'_ LocalCachedRepository)> {
        self.cached_repos_new.iter()
    }

    pub(crate) fn get_repo(&self, path: &Path) -> Option<&LocalCachedRepository> {
        self.cached_repos_new.get(path)
    }
}
